jaq-core/src/filter.rs

```diff
@@ -204,6 +205,26 @@ impl<'a> FilterT<'a> for Ref<'a> {
                     w(if v.as_bool() { then_ } else { else_ }).run(cv.clone())
                 })
             })),
+            Ast::TryCatch(try_, catch_) => {
+                let (ctx, v) = cv;
+                let mut switched = false;
+                let mut stream = w(try_).run((ctx.clone(), v));
+                let step = move || match stream.next() {
+                    None => None,
+                    Some(Err(e)) if switched => Some(Err(e)),
+                    Some(Err(e)) => {
+                        switched = true;
+                        stream = if let Some(catch_filter) = catch_ {
+                            w(catch_filter).run((ctx.clone(), e.as_val()))
+                        } else {
+                            Box::new(std::iter::empty())
+                        };
+                        stream.next()
+                    }
+                    Some(y) => Some(y),
+                };
+                Box::new(std::iter::from_fn(step))
```

Comment:
You could drastically simplify your implementation by following the approach for the pipe operator `Ast::Pipe(l, false, r)`; that is, using `flat_map`.

Reply:
I tried this, but failed in the following way: I need the error emitted inside the `try_` expression to short-circuit its stream. Flat-mapping doesn't allow that (meaning: I couldn't get that to work). That's why I wrote it the way I'd write it with the `stream!` macro and `yield`s: using the closest thing I could find which was `std::iter::from_fn`. I haven't pushed the `flat_map` version of this because of this test:

```text
$ jq -n 'try fma(2,3;1,1;1,"foo",2) catch .'
3
4
3
4
"string (\"foo\") number required"
```

Vs. my flat-mapping version:

```text
$ target/release/jaq.exe -n 'try fma(2,3;1,1;1,"foo",2) catch .'
3.0
"cannot use \"foo\" as float"
4.0
3.0
"cannot use \"foo\" as float"
4.0
4.0
"cannot use \"foo\" as float"
5.0
4.0
"cannot use \"foo\" as float"
5.0
```

Reply:
Another good catch.
How about putting a `take_while` before the `flat_map`?

Reply:
That's almost what I need. `take_while(Result::is_ok)` stops the first stream, but I need the non-ok value to feed into the `catch` filter. How do I do that? My google-fu is lacking.

My first attempt was including the "store the error in a mutable variable" part inside the closure for `take_while`, but then I hit the next wall: for the value to change I'd need to consume the first stream, which is a no-go considering streams might be infinite.

This does look like something that's already implemented in, say, Haskell's Data.List. Something like `take_while_then_chain_feeding_the_first_non_conformant_value(f)`.

Reply:
Oh and there is another reason: d) I consider it an implicit design principle of jq to always use as many values as possible. That is, jq should not simply discard parts of a stream. To me, it feels that the semantics of `?`  and try/catch violate this principle.

Reply:
Furthermore, I would merge Try and TryCatch, because the former is a special case of the latter. For brevity, I would just call the whole thing `Try`.

Reply:
I'll follow your pointers. Still, just to try to justify jq's behaviour, I believe the short-circuiting is what enables this: <https://jqlang.github.io/jq/manual/#breaking-out-of-control-structures>. However, right after introducing the trick, the manual describes the keyword `break` which does the same thing natively.

Reply:
> However, right after introducing the trick, the manual describes the keyword `break` which does the same thing natively.

Yes, and I consider this explicit way of breaking the canonical way to do this example. I wouldn't be opposed, by the way, to implementing label/break, in contrast to implicit shortcutting in try/catch. But I believe that it will be a bit tricky.